src/policy/policy.cpp

```diff
@@ -164,7 +166,7 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
 
         std::vector<std::vector<unsigned char> > vSolutions;
         TxoutType whichType = Solver(prev.scriptPubKey, vSolutions);
-        if (whichType == TxoutType::NONSTANDARD) {
+        if (whichType == TxoutType::NONSTANDARD || whichType == TxoutType::WITNESS_UNKNOWN) {
```

Comment:
This new check seems to overrule the SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_WITNESS_PROGRAM one in VerifyWitnessProgram() ? AFAICT, we don't have test coverage for this standard script flag.

```
                                   AssertionError: [node 1] Expected messages "['non-mandatory-script-verify-flag (Witness version reserved for soft-fork upgrades)']" does not partially match log:
                                    - 2020-08-01T00:14:31.620579Z [msghand] received: tx (87 bytes) peer=1
                                    - 2020-08-01T00:14:31.620845Z [msghand] 52ba45c4ff1da3119475d9ac8201c4ea7018be75444de05aeda9dba22cde28d6 from peer=1 was not accepted: bad-txns-nonstandard-inputs
                                    - 2020-08-01T00:14:31.661127Z [msghand] received: ping (8 bytes) peer=1
```

See new test on both master/diff: https://github.com/ariard/bitcoin/commit/a97e4635ae592e582d186079911d943c58df0e54

Reply:
I'm not sure I understand whether you're asking for a change, or pointing out a bug, or something else?  Could you explain if anything needs to be done here?

Reply:
Ah sorry, previously a segwit v1+ spend would have been rejected by this check : https://github.com/bitcoin/bitcoin/blob/34eb2362581d4d8f0bfd3baa12ba750afaf85c62/src/script/interpreter.cpp#L1552 in `VerifyWitnessProgram`

Now, a segwit v+1 spend is going to be rejected by this new check in `AreInputsStandards`. As linked test confirms if you run it on both master/diff, it's duplicating a policy rule enforcement.

Is this the first case where we duplicate policy rule checks ? I don't know, but we should avoid issues in case of future changes making them inconsistent (like verifying first with `bitcoinconsensus_verify_script` then broadcasting through ATMP and obtaining divergent results). 

If you think it's worthy to fix, we can a) remove L1552 in script/interpreter.cpp or b) add a check against a common `MAX_STANDARD_SEGWIT_VERSION in L1552 or c) just add a comment around both about duplication ?

Reply:
> Is this the first case where we duplicate policy rule checks ?

First, it's not an identical check, because the `AreInputsStandard()` check is only enabled on mainnet and not testnet, via the `fRequireStandard` flag also needing to be set. Whereas the script flag is enabled on all networks. 

Second, I think I'd be pretty surprised if anyone has a mental model of validation where they were certain that policy rule enforcement is not potentially duplicated: the validation rules are so complex and the checks so verbose that we generally tend to take a belt-and-suspenders approach, rather than trying to strictly ensure that any given check only happens once.  If anyone has an understanding that any given transaction policy failure can only happen once in our code, I'd be curious to learn more about how we can be certain of this!

So this doesn't strike me as problematic; moreover I think trying to change validation code in order to ensure there is no duplication is risky and has no benefit.  So I've instead added a comment pointing out that the check might be duplicative.